go/tools/builders/cgo.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2017 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// cgo invokes "go tool cgo" after filtering souces, and then process the output
+// into a normalised form. It is invoked by the Go rules as an action.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/build"
+	"go/parser"
+	"go/token"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+func testCgo(src string, data []byte) (bool, string, error) {
+	fset := token.NewFileSet()
+	parsed, err := parser.ParseFile(fset, src, data, parser.ImportsOnly)
+	if err != nil {
+		return false, "", err
+	}
+	for _, decl := range parsed.Decls {
+		d, ok := decl.(*ast.GenDecl)
+		if !ok {
+			continue
+		}
+		for _, dspec := range d.Specs {
+			spec, ok := dspec.(*ast.ImportSpec)
+			if !ok {
+				continue
+			}
+			imp, err := strconv.Unquote(spec.Path.Value)
+			if err != nil {
+				log.Panicf("%s: invalid string `%s`", src, spec.Path.Value)
+			}
+			if imp == "C" {
+				return true, parsed.Name.String(), nil
+			}
+		}
+	}
+	return false, parsed.Name.String(), nil
+}
+
+func run(args []string) error {
+	sources := multiFlag{}
+	cc := ""
+	objdir := ""
+	flags := flag.NewFlagSet("cgo", flag.ContinueOnError)
+	flags.Var(&sources, "src", "A source file to be filtered and compiled")
+	flags.StringVar(&cc, "cc", "", "Sets the c compiler to use")
+	flags.StringVar(&objdir, "objdir", "", "The output directory")
+	// process the args
+	if len(args) < 2 {
+		flags.Usage()
+		return fmt.Errorf("The go tool must be specified")
+	}
+	gotool := args[0]
+	if err := flags.Parse(args[1:]); err != nil {
+		return err
+	}
+
+	// apply build constraints to the source list
+	// also pick out the cgo sources
+	bctx := build.Default
+	bctx.CgoEnabled = true
+	cgo_srcs := []string{}
```

Comment:
Done, bouncing between skylark and go I think I am going to make a lot of these mistakes :)